src/hooks/questionary/usePersistQuestionaryEditorModel.ts

```diff
@@ -224,31 +203,76 @@ export function usePersistQuestionaryEditorModel() {
             step => step.topic.id === extendedTopicId
           );
 
-          executeAndMonitorCall(() =>
-            assignQuestionsToTopic(
-              state.templateId,
-              reducedTopic!.topic.id,
-              reducedTopic!.fields.map(
-                field => field.question.proposalQuestionId
-              )
-            )
-          );
+          let destinationTopic = reducedTopic;
+
           if (reducedTopicId !== extendedTopicId) {
-            executeAndMonitorCall(() =>
-              assignQuestionsToTopic(
-                state.templateId,
-                extendedTopic!.topic.id,
-                extendedTopic!.fields.map(
-                  field => field.question.proposalQuestionId
-                )
-              )
-            );
+            destinationTopic = extendedTopic;
           }
+
+          const questionRelToChange =
+            destinationTopic?.fields[action.payload.destination.index];
+          const prevQuestion =
+            destinationTopic?.fields[action.payload.destination.index - 1];
+          const nextQuestion =
+            destinationTopic?.fields[action.payload.destination.index + 1];
+
+          const newSortOrder = randomNumberBetween(
+            prevQuestion?.sortOrder,
+            nextQuestion?.sortOrder
+          );
+
+          const questionRel = {
+            ...questionRelToChange,
+            sortOrder: newSortOrder,
+            topicId: destinationTopic?.topic.id,
+          } as QuestionTemplateRelation;
+
+          executeAndMonitorCall(() =>
+            updateQuestionTopicRelation(state.templateId, questionRel)
+          );
           break;
-        case EventType.REORDER_TOPIC_REQUESTED:
-          const topicOrder = state.steps.map(step => step.topic.id);
-          executeAndMonitorCall(() => updateTopicOrder(topicOrder));
+        case EventType.REORDER_TOPIC_REQUESTED: {
+          const sourceIndex = action.payload.source.index;
+          const destinationIndex = action.payload.destination.index;
```

Comment:
These shorter variable names could be used again in later lines instead of the full `action.payload...`